File: PaymentGateway/C#/PGRequestTester.cs
using System;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;
using System.Web;
using System.Xml.Linq;

namespace TestRequests
{
    internal class PgRequestTester
    {
        private HttpClient _httpClient;

        public PgRequestTester()
        {
            _httpClient = new HttpClient();
        }
		
		public static async Task Main(string[] args)
		{
			await MakePostRequestAsync();
			await MakeGetQueryTransactionIdRequestAsync();
		}

        public async Task MakePostRequestAsync()
        {
            string messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><payment><paymentmethod>SVEACARDPAY</paymentmethod><currency>SEK</currency><amount>500</amount><vat>100</vat><customerrefno>125123123</customerrefno><returnurl>https://webpaypaymentgatewaystage.svea.com/webpay-admin/admin/merchantresponsetest.xhtml</returnurl><lang>en</lang></payment>";

            string encodedMessage = Convert.ToBase64String(Encoding.UTF8.GetBytes(messageXML));
            string secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d";
            string mac = GetSha512Hash(encodedMessage + secret);

            Console.WriteLine("mac: " + mac);
            Console.WriteLine("encodedMessage: " + encodedMessage);

            try
            {
                var client = new HttpClient();
                var url = "https://webpaypaymentgatewaystage.svea.com/webpay/payment";

                var content = new FormUrlEncodedContent(new[]
                {
                new KeyValuePair<string, string>("merchantid", "1200"),
                new KeyValuePair<string, string>("message", encodedMessage),
                new KeyValuePair<string, string>("mac", mac)
            });

                HttpResponseMessage response = await _httpClient.PostAsync(url, content);
                string responseString = await response.Content.ReadAsStringAsync();
                Console.WriteLine(responseString);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }

        public async Task MakeGetQueryTransactionIdRequestAsync()
        {
            //string messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><query><customerrefno>temping1238u96896</customerrefno></query>";
            string messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><query><transactionid>900497</transactionid></query>";

            string encodedMessage = Convert.ToBase64String(Encoding.UTF8.GetBytes(messageXML));
            string secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d";
            string mac = GetSha512Hash(encodedMessage + secret);

            Console.WriteLine("mac: " + mac);
            Console.WriteLine("encodedMessage: " + encodedMessage);

            try
            {
                var client = new HttpClient();
                //var url = "https://webpaypaymentgatewaytest.svea.com/webpay/rest/querycustomerrefno";
                var url = "https://webpaypaymentgatewaystage.svea.com/webpay/rest/querytransactionid";

                var content = new FormUrlEncodedContent(new[]
                {
                new KeyValuePair<string, string>("merchantid", "1200"),
                new KeyValuePair<string, string>("message", encodedMessage),
                new KeyValuePair<string, string>("mac", mac)
            });

                HttpResponseMessage response = await _httpClient.PostAsync(url, content);
                string responseString = await response.Content.ReadAsStringAsync();

                Console.WriteLine("Response status code: " + response.StatusCode);
                Console.WriteLine("Response message: " + responseString);

                // Parsing the XML to find the 'message' element
                XDocument xmlDoc = XDocument.Parse(responseString);
                XElement messageElement = xmlDoc.Root.Element("message");
                if (messageElement != null)
                {
                    string encodedMessagePart = messageElement.Value;

                    // Decoding the Base64 message part
                    byte[] decodedBytes = Convert.FromBase64String(encodedMessagePart);
                    string decodedString = Encoding.UTF8.GetString(decodedBytes);
                    Console.WriteLine("Decoded message:");
                    Console.WriteLine(decodedString);
                }
                else
                {
                    Console.WriteLine("Message element not found in response");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }

        private string GetSha512Hash(string input)
        {
            using (SHA512 sha512Hash = SHA512.Create())
            {
                byte[] bytes = sha512Hash.ComputeHash(Encoding.UTF8.GetBytes(input));

                StringBuilder builder = new StringBuilder();
                for (int i = 0; i < bytes.Length; i++)
                {
                    builder.Append(bytes[i].ToString("x2"));
                }
                return builder.ToString();
            }
        }
    }
}

File: PaymentGateway/Php/PGRequestTester.php
<?php

class PgRequest
{
    private $httpClient;

    public function __construct()
    {
        $this->httpClient = curl_init();
    }

    public function makePostRequestAsync()
    {
        $messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><payment><paymentmethod>SVEACARDPAY</paymentmethod><currency>SEK</currency><amount>500</amount><vat>100</vat><customerrefno>125123123</customerrefno><returnurl>https://webpaypaymentgatewaystage.svea.com/webpay-admin/admin/merchantresponsetest.xhtml</returnurl><lang>en</lang></payment>";
        $encodedMessage = base64_encode($messageXML);
        $secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d";
        $mac = $this->getSha512Hash($encodedMessage . $secret);

        echo "mac: " . $mac . "\n";
        echo "encodedMessage: " . $encodedMessage . "\n";

        try {
            $url = "https://webpaypaymentgatewaystage.svea.com/webpay/payment";
            $content = array(
                "merchantid" => "1200",
                "message" => $encodedMessage,
                "mac" => $mac
            );

            $response = $this->makeHttpRequest($url, $content);
            echo "Response status code: " . $response['status_code'] . "\n";
            echo "Response message: " . $response['body'] . "\n";
        } catch (Exception $e) {
            echo $e->getMessage();
        }
    }

    public function makeGetQueryTransactionIdRequestAsync()
    {
        $messageXML = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><query><transactionid>900497</transactionid></query>";
        $encodedMessage = base64_encode($messageXML);
        $secret = "27f18bfcbe4d7f39971cb3460fbe7234a82fb48f985cf22a068fa1a685fe7e6f93c7d0d92fee4e8fd7dc0c9f11e2507300e675220ee85679afa681407ee2416d";
        $mac = $this->getSha512Hash($encodedMessage . $secret);

        echo "mac: " . $mac . "\n";
        echo "encodedMessage: " . $encodedMessage . "\n";

        try {
            $url = "https://webpaypaymentgatewaystage.svea.com/webpay/rest/querytransactionid";
            $content = array(
                "merchantid" => "1200",
                "message" => $encodedMessage,
                "mac" => $mac
            );

            $response = $this->makeHttpRequest($url, $content);
            echo "Response status code: " . $response['status_code'] . "\n";
            echo "Response message: " . $response['body'] . "\n";

            // Parsing the XML to find the 'message' element
            $xml = simplexml_load_string($response['body']);
            $messageElement = $xml->xpath('//message');
            if ($messageElement) {
                $encodedMessagePart = (string)$messageElement[0];

                // Decoding the Base64 message part
                $decodedBytes = base64_decode($encodedMessagePart);
                $decodedString = utf8_decode($decodedBytes);
                echo "Decoded message:\n";
                echo $decodedString . "\n";
            } else {
                echo "Message element not found in response\n";
            }
        } catch (Exception $e) {
            echo $e->getMessage();
        }
    }

    private function getSha512Hash($input)
    {
        return hash('sha512', $input);
    }

    private function makeHttpRequest($url, $postData)
    {
        curl_setopt_array($this->httpClient, array(
            CURLOPT_URL => $url,
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => http_build_query($postData),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HEADER => true,
            CURLOPT_FOLLOWLOCATION => true, // Follow redirects
            CURLOPT_MAXREDIRS => 10 // Maximum number of redirects to follow
        ));

        $response = curl_exec($this->httpClient);
        if (curl_errno($this->httpClient)) {
            throw new Exception(curl_error($this->httpClient));
        }

        $headerSize = curl_getinfo($this->httpClient, CURLINFO_HEADER_SIZE);
        $header = substr($response, 0, $headerSize);
        $body = substr($response, $headerSize);
        $statusCode = curl_getinfo($this->httpClient, CURLINFO_HTTP_CODE);

        return array(
            'status_code' => $statusCode,
            'header' => $header,
            'body' => $body
        );
    }
}

// Usage
$tester = new PgRequest();
$tester->makePostRequestAsync();
$tester->makeGetQueryTransactionIdRequestAsync();

?>
